### ISSUE TITLE
Frenzy problems with bank over 31Sx

### ISSUE THREAD

**Issue by thacerine** [greenc/CookieMaster#82]
If I have a bank over 31Sx cookies, an Elder Frenzy will immediately cause the game to freeze/crash.  In addition, if my bank is below 31Sx at the start of the frenzy, the frenzy will take me up to 31Sx and then freeze. This has been going on for at least a month now, so it's not something caused by one of the most recent updates.

In addition, with the most recent update I encountered another crash problem.  I had a x7 Frenzy running, and a x777 Click Frenzy came up. This immediately caused the game to freeze/crash, even though my bank was not above 31Sx at the time (I believe it was below 10Sx).

**Comment by rashfael** [comment]
Thx for reporting, I'll try to replicate the bug.

**Comment by rashfael** [comment]
Could replicate that, but I have no idea why this is happening. I had autoclicking for the big cookie enabled at the time, did you activate that too?

The tab uses a whole CPU core and nothing in the Developer Tools has any effect.

**Comment by thacerine** [comment]
Yes, I have autoclicking active as well. After running the game a bit more, I've discovered that the freeze after 31Sx now happens with any frenzy at all (even just the regular x7). Before the last set of updates, it was only happening with the Elder Frenzy.

And if more precision helps at all, it appears to be 31.020 Sx that kills it.

**Comment by Nvidiot** [comment]
Extra information: this happens also when auto-clicking the big cookie is disabled (but auto-clicking popups is on).

**Comment by rashfael** [comment]
I did a hard reset and it seems this only happens when you have Heavenly Chips.

**Comment by rashfael** [comment]
After trying to find the bug by provoking it, I used a different approach:

What can possibly bring a tab to 100% CPU and block it completely?

A while loop that never exits. And that is what I looked for. And I found it. On line 1455 of the main.js.

It is also related to javascript Number weirdness. When they get too big, comparison operators return always true.

**Comment by rashfael** [comment]
I am "fixing" this now by limiting that for loop, the "maximum cookie chain award" stat will then be always displayed as maximally 77.777Qi.
Closing it for now, but a better solution is of course welcome.

**Comment by jwelshinger** [comment]
The problem is that when javascript converts a number to a string that is longer than 20 (or is it 21?) it converts it to scientific notation.  So in the line that concatenates the 7 on the end it actually ends up as "7.777777777777778e+217".  In any case parseInt doesn't appear to work with scientific notation, even if that wasn't a ridiculously big number.  By time you reach 17 digits the one's place in decimal has already fallen out of significance range for the mantissa of the floating point, so you will get the exact same results at that point by just multiplying by 10.  The following code could work without an artificial limit on the displayed chain reward:

while(chainValue < bankLimit && chainValue <= cpsLimit) {
          if(chainValue>7e19) {
                 chainValue *= 10;
          } else {
                 chainValue += wrath ? '6' : '7';
                 chainValue = parseInt(chainValue, 10);
          }
      }

**Comment by rashfael** [comment]
Ah, nice. I guessed it was a problem with the comparisons in the loop condition crapping out on large float numbers.
The whole thing could just be replaced by 
`chainValue*10+7`
I can't think of any good reason why this wouldn't work, since the parseInt approach does not fix any problems with floating point precision.

**Comment by jwelshinger** [comment]
The comparison was fine, parseInt was setting chainValue to 7 when the string representation switched to scientific notation (maybe it stops parsing as soon as it hits the decimal point), so chainValue could never get bigger than that value (~7.78e20), it would just keep building up to that value and then resetting to 7 before hitting the bank limit.

Yeah, I agree chainValue*10+7 is the best idea.
